### PR TITLE
SAK-48860 Rubrics ToolItemRubricAssociation improved save method adhering to its equals and hashcode strategy

### PR DESCRIPTION
**Pull request by ern**
https://sakaiproject.atlassian.net/browse/SAK-48860

### CODE REVIEW

File: rubrics/impl/src/main/java/org/sakaiproject/rubrics/impl/repository/AssociationRepositoryImpl.java [hunk 1]
```diff
@@ -97,24 +94,27 @@ public List<ToolItemRubricAssociation> findByItemIdPrefix(String toolId, String
...
+    public ToolItemRubricAssociation save(ToolItemRubricAssociation association) {
+        ToolItemRubricAssociation mergedAssociation = null;
 
-        rubric = rubricRepository.save(rubric);
-        if (index != -1) {
-            return rubric.getAssociations().get(index);
+        Long associationId = association.getId();
+        if (associationId != null) {
+            // updating an existing association we can merge it and return it as nothing changes on the rubric
+            mergedAssociation = (ToolItemRubricAssociation) sessionFactory.getCurrentSession().merge(association);
```
Comment (adrianfish) [inline rubrics/impl/src/main/java/org/sakaiproject/rubrics/impl/repository/AssociationRepositoryImpl.java:103]: I'm not sure I agree with this, tbh. This means that the state isn't consistent until the rubric is loaded  from the db once more, so if anybody in the future uses the rubric's list of associations, they may have a stale association in that list. Isn't it cleaner to just update it in the rubric and let the cascade handle the merge?

File: rubrics/impl/src/main/java/org/sakaiproject/rubrics/impl/repository/AssociationRepositoryImpl.java [hunk 2]
```diff
@@ -97,24 +94,27 @@ public List<ToolItemRubricAssociation> findByItemIdPrefix(String toolId, String
...
-            return rubric.getAssociations().get(rubric.getAssociations().size() - 1);
+            // adding a new association
+            if (association.getRubric() != null) {
+                Long rubricId = association.getRubric().getId();
+                Rubric rubric = rubricRepository.getById(rubricId);
+                if (rubric != null) {
+                    rubric.getAssociations().add(association);
+                    rubricRepository.save(rubric);
+                    // query for the new association
+                    mergedAssociation = findByItemIdAndRubricId(association.getItemId(), rubricId).orElse(null);
```
Comment (adrianfish) [inline rubrics/impl/src/main/java/org/sakaiproject/rubrics/impl/repository/AssociationRepositoryImpl.java:113]: This line implies that a null can be returned. I know it probably never will be, but the code reads like that.

File: rubrics/impl/src/main/java/org/sakaiproject/rubrics/impl/repository/AssociationRepositoryImpl.java (hunk 1 above)
Comment (ern) [reply rubrics/impl/src/main/java/org/sakaiproject/rubrics/impl/repository/AssociationRepositoryImpl.java:103]: Note that the association was simply updated hibernate will internally update it's references to the updated association so no harm there and no the list of associations would not be stale. We absolutely could let the merge on the rubric cascade the update to the association this is simply faster and it achieves the same goal. Now if it were adjusting the actual list of associations then you would take the approach that you mention which it does, in the subsequent lines after this.

Comment (ern) [reply rubrics/impl/src/main/java/org/sakaiproject/rubrics/impl/repository/AssociationRepositoryImpl.java:103]: You are correct though in noticing this shortcut.

File: rubrics/impl/src/main/java/org/sakaiproject/rubrics/impl/repository/AssociationRepositoryImpl.java (hunk 2 above)
Comment (ern) [reply rubrics/impl/src/main/java/org/sakaiproject/rubrics/impl/repository/AssociationRepositoryImpl.java:113]: Correct a null is definitely possible to be returned, there are checks to ensure the association has a rubric. Should an association be persisted without a rubric? maybe the correct thing is to raise an exception here vs returning null?

File: rubrics/impl/src/main/java/org/sakaiproject/rubrics/impl/repository/AssociationRepositoryImpl.java [hunk 3]
```diff
@@ -97,24 +94,27 @@ public List<ToolItemRubricAssociation> findByItemIdPrefix(String toolId, String
...
+            // updating an existing association we can merge it and return it as nothing changes on the rubric
+            mergedAssociation = (ToolItemRubricAssociation) sessionFactory.getCurrentSession().merge(association);
         } else {
-            return rubric.getAssociations().get(rubric.getAssociations().size() - 1);
+            // adding a new association
+            if (association.getRubric() != null) {
+                Long rubricId = association.getRubric().getId();
+                Rubric rubric = rubricRepository.getById(rubricId);
+                if (rubric != null) {
+                    rubric.getAssociations().add(association);
```
Comment (ern) [inline rubrics/impl/src/main/java/org/sakaiproject/rubrics/impl/repository/AssociationRepositoryImpl.java:110]:
```suggestion
                    association.setRubric(rubric);
                    rubric.getAssociations().add(association);
```

**Review by adrianfish — APPROVED** [review]
Looks good.

**Pull request merged**